### PR TITLE
Validate follower pod owned by same Job as leader pod

### PR DESCRIPTION
**Pull request by danielvegamyhre**
This change validates the leader pod has same owner UID as the follower, to ensure they are part of the same Job.
This is necessary to handle a potential race condition between index updates and pod rescheduling during JobSet restarts.

1. A job failure occurs and the JobSet is restarted (deleting and recreating all jobs)
2. Leader pods may land on different node pools than they were originally scheduled on.
3. When the follower pods are recreated, and we look up the leader pod using the index which maps
[pod name without random suffix] -> corev1.Pod object, if this occurs before the index updates for the leader pod have been pushed to the controller, we may get a stale index entry and inject the the wrong nodeSelector, using the topology the leader pod was originally scheduled on before the
restart.

### CODE REVIEW

File: pkg/webhooks/pod_admission_webhook.go [hunk 1]
```diff
@@ -82,7 +83,7 @@ func (p *podWebhook) leaderPodScheduled(ctx context.Context, pod *corev1.Pod) (b
 	}
 	scheduled := leaderPod.Spec.NodeName != ""
 	if !scheduled {
-		log.V(3).Info("leader pod %s is not yet scheduled", leaderPod.Name)
+		log.V(0).Info("leader pod %s is not yet scheduled", leaderPod.Name)
```
Comment (kannon92) [inline pkg/webhooks/pod_admission_webhook.go:86]:
```suggestion
		log.V(2).Info("leader pod %s is not yet scheduled", leaderPod.Name)
```

Either we should use level 2 or just remove the V(0) and use log.Info(...)

What log level do we deploy jobset with? I think 2 is usually default logging anyway.

File: pkg/webhooks/pod_admission_webhook.go [hunk 2]
```diff
@@ -131,3 +142,20 @@ func genLeaderPodName(pod *corev1.Pod) (string, error) {
 	leaderPodName := placement.GenPodName(jobSet, replicatedJob, jobIndex, "0")
 	return leaderPodName, nil
 }
+
+// validatePodsOnSameRestartAttempt returns an error if the leader pod and
+// follower pod are not owned by the same Job UID. Otherwise, it returns nil.
+func podsOwnedBySameJob(leaderPod, followerPod *corev1.Pod) error {
```
Comment (kannon92) [inline pkg/webhooks/pod_admission_webhook.go:148]: Could you add some tests for this?

**Comment by kannon92** [comment]
I am going to leave LGTM for @ahg-g on this one. I don't really have much context into this problem.

File: pkg/webhooks/pod_admission_webhook.go (hunk 1 above)
Comment (ahg-g) [reply pkg/webhooks/pod_admission_webhook.go:86]: yeah, I think we generally have been using 2 I think

File: pkg/webhooks/pod_admission_webhook.go (hunk 2 above)
Comment (ahg-g) [reply pkg/webhooks/pod_admission_webhook.go:148]: +1

Comment (danielvegamyhre) [reply pkg/webhooks/pod_admission_webhook.go:148]: Added unit tests

File: pkg/webhooks/pod_admission_webhook.go (hunk 1 above)
Comment (danielvegamyhre) [reply pkg/webhooks/pod_admission_webhook.go]: Done

**Pull request merged**